src/components/input/input.ts

```diff
@@ -0,0 +1,108 @@
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ */
+const growElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Grow until we stop scrolling
+  while (isScrolling) {
+    rows += 1;
+    textarea.setAttribute('rows', String(rows));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+  }
+};
+
+/**
+ * Shrink Elastic TextArea
+ *
+ * Decreases the `rows` attribute of a `textarea` until it begins scrolling.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ */
+const shrinkElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Shrink until we hit the minimum rows or start scrolling
+  for (let i = rows; i >= minRows; i--) {
+    rows = i;
+    textarea.setAttribute('rows', String(Math.max(rows, minRows)));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+    // If we're scrolling, then we shrank too much, so grow a bit and stop
+    if (isScrolling) {
+      growElasticTextArea(textarea);
+      break;
+    }
+  }
+};
+
+/**
+ * Update Elastic TextArea
+ *
+ * When a textarea is updated, compares the new value to the previous value
+ * (from a dataset attribute) to determine whether to grow or shrink.
+ *
+ * Note: Ideally we'd destructure `event.target` from the parameters and declare
+ * the type directly, since we don't care about the rest of the `event` object.
+ * Unfortunately, that caused some TS problems. If someone finds a solution,
+ * we should update this function to take `{target}` as the only parameter.
+ * @see https://stackoverflow.com/a/28900856/4898045
+ *
+ * @param {Event} event - the input event on the target `textarea`
+ */
+const updateElasticTextArea = (event: Event) => {
+  const textarea = event.target as HTMLTextAreaElement;
+  const lastValue = textarea.dataset.lastValue || '';
+  const newValue = textarea.value;
+
+  // Check whether to grow or shrink
+  if (newValue.length > lastValue.length) {
+    growElasticTextArea(textarea);
+  } else {
+    shrinkElasticTextArea(textarea);
+  }
+
+  // Save the new value for comparing
+  textarea.dataset.lastValue = newValue;
+};
+
+/**
+ * Create Elastic TextArea
+ *
+ * Given a textarea, adds dataset attributes for the existing value and the
+ * minimum number of rows, for use in the update function, which is added to
+ * an event listener. Returns an object containing the `destroy()` method to
+ * remove the event listener.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ * @returns {object} object containing a `destroy()` method
```

Comment:
@calebeby Is there a better way to document what is returned? What happens when we are returning an object that has multiple methods? It seems trying to describe every method/function that is returned would turn into a super-long sentence. 😬

Reply:
In the past I've followed this strategy for defining a `param` option with multiple props. I wonder if we could do something similar for returns?

https://stackoverflow.com/a/6460748/7816145

Reply:
Though I guess the returned object is unnamed... 🤔

Reply:
If it was me, I would remove the entire `@returns` annotation. TS already knows what this returns. If you hover over the `=>`, it will show the signature, which includes the (inferred) return type.

For me, describing the return type in words does not help me understand the function.

I would add a description annotation to the `destroy` method, which would be helpful:

![image](https://user-images.githubusercontent.com/13206945/82488826-73b0e980-9a95-11ea-9a2e-a5ed74cae153.png)

Reply:
> I would add a description annotation to the destroy method, which would be helpful:

I agree! 😄